cpp/ql/src/experimental/Security/CWE/CWE-552/DangerousUseRename.ql

```diff
@@ -0,0 +1,58 @@
+/**
+ * @name Dangerous use rename.
+ * @description If the rename function did not work correctly, simply overwriting the destination file can lead to errors.
+ * @kind problem
+ * @id cpp/dangerous-use-rename
+ * @problem.severity warning
+ * @precision medium
+ * @tags correctness
+ *       security
+ *       external/cwe/cwe-552
+ */
+
+import cpp
+import semmle.code.cpp.valuenumbering.GlobalValueNumbering
+
+/** Holds if the function argument is used in opening the file for reading or writing. */
+predicate findFileForReadOrWrite(FunctionCall fc, FunctionCall fc1, string mode, int na) {
+  fc1.getTarget().hasName(["open", "fopen"]) and
+  globalValueNumber(fc1.getArgument(0)) = globalValueNumber(fc.getArgument(na)) and
+  fc1.getArgument(1).getValue() = mode and
+  fc.getASuccessor+() = fc1
+}
+
+from FunctionCall fc
+where
+  fc.getTarget().hasName("rename") and
+  exists(IfStmt ifst, Expr ec, Expr ecd, FunctionCall fc1, FunctionCall fc2 |
+    findFileForReadOrWrite(fc, fc1, "r", 0) and
```

Comment:
Why are you interested in `open(from, "r")`? I see this back in all the examples/tests, but the text output by `select` does not make mention of this and effectively only refers to `open(to, "w")`

Reply:
this is necessary to detect exactly the manual renaming of files when a source file and a target file are encountered and the data is copied.

Reply:
Right, so are you interested in situations where, first, the renaming fails, and then the source and target files are opened and the contents is copied manually. If so, *all* the comments and qhelp need to explain this, because at the moment the impression I get from the qhelp is that you're just interested in failing renames followed by opening of the target file.